src/measure_theory/group/add_circle.lean

```diff
@@ -0,0 +1,107 @@
+/-
+Copyright (c) 2022 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import measure_theory.integral.periodic
+import data.zmod.quotient
+
+/-!
+# Measure-theoretic results about the additive circle
+
+The file is a place to collect measure-theoretic results about the additive circle.
+
+## Main definitions:
+
+ * `add_circle.closed_ball_ae_eq_ball`: open and closed balls in the additive circle are almost
+   equal
+ * `add_circle.is_add_fundamental_domain_of_ae_ball`: a ball is a fundamental domain for rational
+   angle rotation in the additive circle
+
+-/
+
+open set function filter measure_theory measure_theory.measure metric
+open_locale measure_theory pointwise big_operators topological_space ennreal
+
+namespace add_circle
+
+variables {T : ℝ} [hT : fact (0 < T)]
+include hT
+
+lemma closed_ball_ae_eq_ball {x : add_circle T} {ε : ℝ} :
```

Comment:
You shouldn't need `hT` for this one.

Reply:
Agreed. However unfortunately `add_circle.measure_space` requires `hT` so I'd have to change that instance to implement this.

I'm happy to do this if you think I should?

Reply:
Indeed, `hT` should not be needed to define the measure on the quotient (we're quotienting the real line by a discrete subgroup, so the measure on the quotient always makes sense). Ideally, we shouldn't even need to discuss whether T is zero or not in the definition of the measure...

Reply:
Hmm, I don't see an easy way to avoid branching on whether `T` is zero (though I agree this is desirable).

I suppose one approach is to change [this definition](https://github.com/leanprover-community/mathlib/blob/3891ab62ecddc54174f19188fd850a5d6f3cfdde/src/measure_theory/integral/periodic.lean#L59) to:
```lean
noncomputable instance measure_space : measure_space (add_circle T) :=
{ volume := (if T = 0 then volume else volume.restrict $ Ioc 0 (|T|)).map coe,
  .. add_circle.measurable_space }
```
and then prove:
```lean
variables [hT : fact (0 < T)]
include hT

/-- When `0 < T`, the volume on `add_circle T` is just `T` times the Haar measure. -/
lemma volume_eq : (volume : measure $ add_circle T) = ennreal.of_real T • add_haar_measure ⊤ :=
begin
  change (if T = 0 then volume else volume.restrict $ Ioc 0 (|T|)).map coe = _,
  simp only [hT.out.ne.symm, if_false, abs_eq_self.mpr hT.out.le],
  letI : measurable_space (ℝ ⧸ zmultiples T) := add_circle.measurable_space,
  letI : borel_space (ℝ ⧸ zmultiples T) := add_circle.borel_space,
  letI : compact_space (ℝ ⧸ zmultiples T) := add_circle.compact_space T,
  have := (is_add_fundamental_domain_Ioc' hT.out 0).map_restrict_quotient ⊤ (by simp),
  simp only [zero_add] at this,
  erw this,
  simp,
end
```
and proceed from there.

What do you think?

Reply:
This is turning out to be non-trivial because quite a few facts for the circle are proved for `0 < T` rather than `T ≠ 0`. In several cases, we would need their generalisation not just for proofs but even for statements. For example We only know the circle is compact when `0 < T` and so we cannot talk about `add_haar_measure ⊤` when `T <  0`.

After experimenting with a few approaches I think the only clean way to do things would be to replace the `fact (0 < T)` pattern entirely with `fact (T ≠ 0)`. (I admit a bias here -- I did not actually introduce the `fact (0 < T)` pattern and have always harboured some reservations). I am willing to do this but I would like to make the case that it should happen in a follow-up PR (which would then also address the point here about `closed_ball_ae_eq_ball`).

I have considered alternative approaches including ingredients such as:

- allowing a mixture of `fact (0 < T)` and `fact (T ≠ 0)` hypotheses
- attempt to stick with `fact (0 < T)` throughout and arguing using a measure-preserving map `add_circle T → add_circle (-T)`
- directly attack the problem that we have two names for the same subgroup (`zmultiples T`, `zmultiples (-T)`) and thus two different types `add_circle T` and `add_circle (-T)`. E.g., redefining `add_circle` to consume a subgroup together with a discreteness hypotheses or a proof that it is set-wise `zmultiples T` for some `T`

I have mostly convinced myself that these really just push the mess around and the best solution is to migrate to `fact (T ≠ 0)` but I have fully convinced myself that it is reasonable for this work to take place in a separate PR. (Nevertheless, I welcome dissenting opinions.)

What do you think?

Reply:
I agree that this is better for a followup PR. You can merge this one once #17590 is in.
bors d+